delira/training/predictor.py

```diff
@@ -0,0 +1,208 @@
+
+                batch_dict = {}
+                for batch in batch_list:
+                    for key, val in batch.items():
+                        if key in batch_dict.keys():
+                            batch_dict[key].append(val)
+                        else:
+                            batch_dict[key] = [val]
+
+                for key, val_list in batch_dict.items():
+                    batch_dict[key] = np.array(val_list)
+
+                preds = self.predict(batch_dict)
+
+                predictions_all.append(preds)
+                inputs_all.append(batch_dict)
+
+                batch_list = []
+
+        batchgen._finish()
+
+        predictions_all = zip(*predictions_all)
+        predictions_all = [np.vstack(_outputs) for _outputs in predictions_all]
+
+        input_keys = inputs_all[0].keys()
+
+        total_input_dict = {}
+
+        for key in input_keys:
+            total_input_dict[key] = np.concatenate(
+                [_input[key] for _input in inputs_all])
+
+        datamgr.batch_size = orig_batch_size
+        datamgr.n_process_augmentation = orig_num_aug_processes
+
+        return predictions_all, total_input_dict
+
+    def __setattr__(self, key, value):
+        """
+        Set attributes and guard specific attributes after they have been set
+        once
+
+        Parameters
+        ----------
+        key : str
+            the attributes name
+        value : Any
+            the value to set
+
+        Raises
+        ------
+        PermissionError
+            If attribute which should be set is guarded
+
+        """
+
+        # check if key has been set once
+        if key in self.__KEYS_TO_GUARD and hasattr(self, key):
+            raise PermissionError("%s should not be overwritten after "
+                                  "it has been set once" % key)
+        else:
+            super().__setattr__(key, value)
+
+    @staticmethod
+    def _is_better_val_scores(old_val_score, new_val_score,
+                              mode='highest'):
+        """
+        Check whether the new val score is better than the old one
+        with respect to the optimization goal
+
+        Parameters
+        ----------
+        old_val_score :
+            old validation score
+        new_val_score :
+            new validation score
+        mode: str
+            String to specify whether a higher or lower validation score is
+            optimal; must be in ['highest', 'lowest']
+
+        Returns
+        -------
+        bool
+            True if new score is better, False otherwise
+        """
+
+        assert mode in ['highest', 'lowest'], "Invalid Comparison Mode"
+
+        if mode == 'highest':
+            return new_val_score > old_val_score
+        elif mode == 'lowest':
+            return new_val_score < old_val_score
+
+    @staticmethod
+    def calc_metrics(groundtruths, *predictions, metrics={}, metric_keys=None):
```

Comment:
`calc_metrics` is called in the network trainer without the option to change the metrics_keys. 

https://github.com/justusschock/delira/blob/722cd55dd308f2ff067f34d5eb323511fe3073c2/delira/training/abstract_trainer.py#L334-L338

It may be useful to introduce a new argument in the trainer to set the metrics_keys so it is not always necessary to overwrite the function if different metric keys are needed.